instance/management/commands/instance_statistics_csv.py

```diff
@@ -0,0 +1,221 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app - Activity report management command
+"""
+
+# Imports #####################################################################
+
+from configparser import ConfigParser
+import csv
+from datetime import datetime
+import os
+import sys
+
+from django.conf import settings
+from django.core.management.base import BaseCommand
+from django.db.models import Q
+
+from instance import ansible
+from instance.models.openedx_instance import OpenEdXInstance
+from registration.models import BetaTestApplication
+
+
+# Classes #####################################################################
+
+
+class Command(BaseCommand):
+    """
+    Logs_activity_csv management command class
+    """
+    help = (
+        'Generates a CSV containing basic activity information about all app servers'
+        ' (numbers for hits, distinct hits, users, and courses).'
+    )
+
+    LOGS_SERVER = 'logs.opencraft.com'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--qualified-domain',
+            '-q',
+            default=None,
+            help='The fully qualified domain name to filter results by',
+            required=True
+        )
+        parser.add_argument(
+            '--days',
+            '-d',
+            default=60,
+            help='The number of days that should be queried for statistics. Default is 60 days'
+        )
+        parser.add_argument(
+            '--out',
+            default=None,
+            help='Path to the output file of the new CSV. Leave blank to use stdout.'
+        )
+
+    def handle(self, *args, **options):
+        # Determine the stream to be used for outputting the CSV.
+        if options['out'] is None:
+            out = self.stdout
+        else:
+            try:
+                out = open(options['out'], 'w')
+            except PermissionError:
+                self.stderr.write(self.style.ERROR(
+                    'Permission denied while attempting to write file: {outfile}'.format(outfile=options['out'])
+                ))
+                sys.exit(1)
+
+        self.collect_instance_statistics(out, options['qualified_domain'], options['days'])
+
+    def get_instance_from_qualified_domain(self, qualified_domain):
+        try:
+            instance = OpenEdXInstance.objects.get(
+                Q(external_lms_domain=qualified_domain) | Q(internal_lms_domain=qualified_domain)
+            )
+        except OpenEdXInstance.DoesNotExist:
+            self.stderr.write(self.style.ERROR(
+                'No OpenEdXInstance exists with an external or internal domain of {qualified_domain}'.format(qualified_domain=qualified_domain)
+            ))
+            sys.exit(1)
+
+        # If there are no appservers for the instance, we should error out
+        if instance.appserver_set.first() is None:
```

Comment:
@pcockwell, this check and the error message might be misleading. You can check the `successfully_provisioned` field instead and check if the instance has active appservers (required for running the playbook).

Reply:
Switched this to:
```
if not instance.successfully_provisioned or not instance.get_active_appservers():
```
This will ensure that it is successfully provisioned and has a non-empty set of active appservers, or it will error.